### ISSUE TITLE
Sync client hang

### ISSUE THREAD

**Issue by marcbrevoort-cyberhive** [bluejekyll/trust-dns#2050]
**Describe the bug**
Using the sync client, the query
```
    let response = client.query(&name, dns_class, record_type);
```
may hang and take forever to complete.

**To Reproduce**
Left a process running overnight.

**Expected behavior**
If response not received within a reasonable time frame, the function
```
    fn query(
        &self,
        name: &Name,
        query_class: DNSClass,
        query_type: RecordType,
    ) -> ClientResult<DnsResponse> {
        let (mut client, runtime) = self.spawn_client()?;

        runtime.block_on(client.query(name.clone(), query_class, query_type))
    }
```
could just return an error or an empty result. Alternatively, since the sync function just seems a wrapper for the async function, a parallel to the async library call `with_timeout` could be provided.

**System:**
 - OS: Windows (observed), probably others
 - Architecture: x86_64
 - Version 0.22.0
 - rustc version: 1.72.1

**Version:**
trust-dns-client = { version = "0.22.0", default-features = false, features = ["dns-over-https", "dns-over-https-rustls"]}

**Additional context**
librustls reports
```[06:39:41.767] (9a90) DEBUG  No cached session for DnsName(DnsName(DnsName("dns.google")))``` - this nested DnsName seems odd.

**Comment by bluejekyll** [comment]
Maybe we need some better defaults here, but could you try setting a timeout with this method when you construct your connection for the Client? https://docs.rs/trust-dns-client/0.23.0/trust_dns_client/udp/struct.UdpClientConnection.html#method.with_timeout

**Comment by marcbrevoort-cyberhive** [comment]
I'll try - but am currently on 0.22.0 using HttpsClientConnection and this no longer seems to exist in 0.23.0 - so this will require a fairly severe overhaul.

**Comment by marcbrevoort-cyberhive** [comment]
Actually the UdpClientConnection will not do - while I can make it do lookups, WireShark shows it is not secured. I have looked for a timeout function in HttpsClientConnection but cannot find it. Is there a way I'm unaware of that allows me to do DNS over HTTPS _and_ also set the maximum timeout to something non-infinite?

**Comment by marcbrevoort-cyberhive** [comment]
Have upgraded to Hickory (0.24) now. I'm still at a loss about how to set a timeout for HttpsClientConnection.
The documentation states 

> Use with hickory_client::client::Client impls

but I don't understand in what way this is intended. Any insights please?

**Comment by bluejekyll** [comment]
Hm, I was just taking a look at the http interface, and you're right, there's not timeout option to pass through on the configuration. This seems like a missing option. I had thought we had an interface like this to control timeouts, https://docs.rs/hickory-client/0.24.0/hickory_client/udp/struct.UdpClientStream.html#method.with_timeout

but it appears not. we should definitely add one.

**Comment by marcbrevoort-cyberhive** [comment]
> We should definitely add [with_timeout].
May I ask if there's a known timeline for this yet?

**Comment by bluejekyll** [comment]
I'm not sure that anyone has stepped forward to offer a patch for this at this point. I was just looking through the code, and while it would be easy to add the TCP parameters here: https://github.com/hickory-dns/hickory-dns/blob/main/crates/proto/src/h2/h2_client_stream.rs#L325C26-L325C43 and probably here: https://github.com/hickory-dns/hickory-dns/blob/main/crates/resolver/src/name_server/connection_provider.rs#L355C22-L355C22

you can see the timeout example in the standard TCP impl here: https://github.com/hickory-dns/hickory-dns/blob/main/crates/resolver/src/name_server/connection_provider.rs#L303

If we want to implement timeouts on individual operations, I think that will take more effort, and I'd recommend that we rewrite the H2 client stream into modern async/await syntax before starting to do that. Also, it looks like we need timeouts on h2, h3, and quic. I think this logic is baked into the tcp and tls impls in the same way with the `DnsMultiplexer` implementation, but since h2, h3, and quic have multiplexing baked in, that interface wasn't necessary for them which is probably why timeouts aren't as robust there right now. That's also probably worth reviewing to see if it offers any good methods for implementing these timeouts.

**Comment by marcbrevoort-cyberhive** [comment]
if I add a timeout param at https://github.com/hickory-dns/hickory-dns/blob/main/crates/proto/src/h2/h2_client_stream.rs#L325C26-L325C43 and recompile, I'm getting no errors or anything? Where do I go next? With apparently no scrutiny from the compiler, it's rather hard to trace how the code fits together.

**Comment by bluejekyll** [comment]
Yes, I think in many cases that is the case. We rely on the socket being dropped to cleanup and close all the connections. I'd have to look at all the implementations to remember how and where we did all of that. A bunch of that is older code that I haven't looked at in a while.

**Comment by marcbrevoort-cyberhive** [comment]
I'm probably oversimplifying, but why not use the built-in timeout functionality, e.g. `tcpstream.set_read_timeout(timeout);` ?

**Comment by marcbrevoort-cyberhive** [comment]
I've ultimately resorted to wrapping the DNS call in a function that shuts down the call if it doesn't return in a timely fashion, broadly following [this example](https://docs.rs/tokio/latest/tokio/runtime/struct.Runtime.html#method.shutdown_timeout) but also allowing the wrapper to store the result in an `Arc<Mutex<Option<Result<DnsResponse, ClientError>>>>`. On timeout, there is no result, and this None is handled by raising a timeout exception. Otherwise the Result<DnsResponse, ClientError> is returned.

The benefit is that it doesn't touch Hickory, and when the update lands, in principle it should just keep working - so both fixing the lack of timeout in Hickory and removing the wrapper once Hickory is fixed can happen in slow time.
